### ISSUE TITLE
Avoid modifying the process global environment (not thread safe)

### ISSUE THREAD

**Issue by gpshead** [python/cpython#83557]
BPO | [39376](https://bugs.python.org/issue39376)
--- | :---
Nosy | @gpshead, @ericsnowcurrently

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-01-18.00:47:44.215>
labels = ['expert-subinterpreters', 'interpreter-core', '3.9', 'type-crash']
title = 'Avoid modifying the process global environment (not thread safe)'
updated_at = <Date 2021-04-16.00:43:07.242>
user = 'https://github.com/gpshead'
```

bugs.python.org fields:
```python
activity = <Date 2021-04-16.00:43:07.242>
actor = 'vstinner'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core', 'Subinterpreters']
creation = <Date 2020-01-18.00:47:44.215>
creator = 'gregory.p.smith'
dependencies = []
files = []
hgrepos = []
issue_num = 39376
keywords = []
message_count = 2.0
messages = ['360222', '360225']
nosy_count = 2.0
nosy_names = ['gregory.p.smith', 'eric.snow']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'needs patch'
status = 'open'
superseder = None
type = 'crash'
url = 'https://bugs.python.org/issue39376'
versions = ['Python 3.9']
```

</p></details>

**Comment by ericsnowcurrently** [comment]
@colesbury, is this something that needs to be addressed for free-threaded builds?

**Comment by gpshead** [comment]
given modifying the C/kernel provided environment blob is already broken regardless of free-threaded builds, I doubt it needs special attention beyond making sure the existing `os.putenv(x, y)` and `os.environ[x] = y` documentation calls it out.

those APIs haven't been compatible with the modern world for decades.

**Comment by gpshead** [comment]
If we went with my idea in the opening message of tracking local modifications, it might be a challenge.  But my gut feeling is that we just won't bother doing this?  I don't recall the last time I've seen this come up as an issue.  _Hopefully_ that means people just don't do it and pass new env= mappings to subprocesses?

**Comment by colesbury** [comment]
I agree with what @gpshead wrote -- I think we should document it, but I don't think there's a good fix for it at the Python level.

The same problem affects other languages and runtimes:
* Go: https://github.com/golang/go/issues/63567#issuecomment-1779931060
* Rust: https://github.com/rust-lang/rust/pull/124636 (marked "unsafe" recently)

Rachel Kroll had a [blog post](https://rachelbythebay.com/w/2017/01/30/env/) ~7 years ago about setenv that concludes with:

> None of this is new, but we do re-discover it roughly every five years.
> See you in 2022.

I expect we'll have users run into this problem every so often, even with updated docs. It's too bad because I think the technical fix for glibc, etc. is not too complex.

**Comment by gpshead** [comment]
While I mostly like to pretend `setenv()` or modifying main's `char **environ` doesn't exist and the world has _largely_ adopted the same mentality...  The code I added recently to try and detect a multithreaded process to warn people who were forking could be generalized and reused here for an environment modification warning when threads exist.

https://github.com/python/cpython/blob/main/Modules/posixmodule.c#L7884 `warn_about_fork_with_threads()`

**Comment by tbu-** [comment]
Permalink:

https://github.com/python/cpython/blob/0890ad7c024ccf29614849b6ffadcb92c0e91ce7/Modules/posixmodule.c#L7884
